src/quick-lint-js/parse.h

```diff
@@ -2516,7 +2516,27 @@ class parser {
 
     if (this->peek().type == token_type::kw_else) {
       this->skip();
+
+      const token token_after_else = this->peek();
       parse_and_visit_body();
+      const token expect_left_curly = this->peek();
+
+      lexer_transaction transaction = this->lexer_.begin_transaction();
+      this->lexer_.skip();
+
+      if (this->peek().type == token_type::right_curly) {
+        this->error_reporter_->report(error_else_has_empty_body{
+            .where = source_code_span(expect_left_curly.begin, this->peek().end)
+        });
+      } else if (token_after_else.type == token_type::semicolon &&
```

Comment:
This is the check for an empty block on the else case. Now that I read it again, I see it should only be triggered if the else has a semicolon 🤦‍♂️
https://github.com/quick-lint/quick-lint-js/pull/290#discussion_r631650388

Is creating a transaction on the lexer and then rolling it back the correct way to peek forward?

Reply:
> Is creating a transaction on the lexer and then rolling it back the correct way to peek forward?

We only need a transaction if you want to look ahead multiple tokens.

Can you give an example JS program where using `this->peek()` won't work?

Reply:
I had gone through a rabbit hole of actually checking the body of the next scope to the else. That's why I was advancing the cursor and parsing the next scope after the else using a transaction 🤦‍♂️ . I've rolled the code back to only check for the newline and the opening brace.
Commits are now squashed, let me know what you think.